File: 406.go
package main

import "sort"

func reconstructQueue(people [][]int) [][]int {
	var highs []int
	cache := make(map[int]bool)
	highSorts := make(map[int][][]int)
	for _, human := range people {
		if _, ok := cache[human[0]]; !ok {
			cache[human[0]] = true
			highs = append(highs, human[0])
		}
		highSorts[human[0]] = append(highSorts[human[0]], human)
	}
	sort.Sort(sort.Reverse(sort.IntSlice(highs)))
	var res [][]int
	for _, high := range highs {
		highSort := highSorts[high]
		sort.SliceStable(highSort, func(i, j int) bool {
			return highSort[i][1] < highSort[j][1]
		})
		for _, human := range highSort {
			res = append(res, []int{})
			copy(res[human[1]+1:], res[human[1]:])
			res[human[1]] = human
		}
	}
	return res
}

File: 406.py
from typing import List


class Solution:
    def reconstructQueue(self, people: List[List[int]]) -> List[List[int]]:
        from collections import defaultdict

        highs, highs_sort = set(), defaultdict(list)
        for human in people:
            highs.add(human[0])
            highs_sort[human[0]].append(human)
        res = []
        for high in reversed(sorted(highs)):
            for human in sorted(highs_sort[high], key=lambda x: x[1]):
                res.insert(human[1], human)
        return res


if __name__ == '__main__':
    s = Solution()
    s.reconstructQueue([[9, 0], [7, 0], [1, 9], [3, 0], [2, 7], [5, 3], [6, 0], [3, 4], [6, 2], [5, 2]])
